contracts/sky/src/handle.rs

```diff
@@ -1,273 +1,475 @@
+            // add the payback msg
+            messages.push(send_msg(
+                env.message.sender,
+                c_std::Uint128(payback_amount.u128()),
+                None,
+                None,
+                None,
+                1,
+                cur_asset.code_hash.clone(),
+                cur_asset.address.clone(),
+            )?);
+        }
+        _ => {}
+    }
+
+    // the final cur_asset should be the same as the start_addr
+    if !(cur_asset.clone() == Cycles::load(&deps.storage)?.0[index.u128() as usize].start_addr) {
+        return Err(StdError::generic_err(
+            "final asset not equal to start asset",
+        ));
     }
 
-    Ok(HandleResponse{
+    Ok(HandleResponse {
         messages,
         log: vec![],
-        data: Some(to_binary(&HandleAnswer::ExecuteArb{
+        data: Some(to_binary(&HandleAnswer::ExecuteArbCycle {
             status: true,
-        })?)
+            swap_amounts: return_swap_amounts,
+            payback_amount,
+        })?),
     })
-}*/
+}
 
-pub fn try_execute<S: Storage, A: Api, Q: Querier>(
+pub fn try_arb_all_cycles<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
     env: Env,
     amount: Uint128,
 ) -> StdResult<HandleResponse> {
-
-    let config: Config = Config::load(&deps.storage)?;
-
-    //if amount.gt(env.)
-    
-    let res = trade_profitability( deps, amount )?;
-
-    let mut profitable = false;
-    let mut is_mint_first = false;
-    let mut pool_shd_amount = Uint128::zero();
-    let mut pool_silk_amount = Uint128::zero();
-    let mut first_swap_min_expected = Uint128::zero();
-    let mut second_swap_min_expected = Uint128::zero();
+    let mut total_profit = Uint128::zero();
+    let mut messages = vec![];
+    let res = any_cycles_profitable(deps, amount)?; // get profitability data from query
     match res {
-        sky::QueryAnswer::TestProfitability{
-            is_profitable, 
-            mint_first, 
-            shd_amount,
-            silk_amount,
-            first_swap_amount, 
-            second_swap_amount,
+        sky::QueryAnswer::IsAnyCycleProfitable {
+            is_profitable,
+            profit,
+            ..
         } => {
-            profitable = is_profitable;
-            is_mint_first = mint_first;
-            pool_shd_amount = shd_amount;
-            pool_silk_amount = silk_amount;
-            first_swap_min_expected = first_swap_amount;
-            second_swap_min_expected = second_swap_amount;
+            // loop through the data returned for each cycle
+            for (i, profit_bool) in is_profitable.iter().enumerate() {
+                // if a cycle is profitable call the try_arb_cycle fn and keep track of the
+                // total_profit
+                if profit_bool.clone() {
+                    messages.push(
+                        sky::HandleMsg::ArbCycle {
+                            amount,
+                            index: Uint128::from(i as u16),
```

Comment:
```suggestion
                            index: Uint128::from(i as u128),
```

using as u16 makes the struct later convert into u128

Reply:
Done